### PR TITLE
Add unary get support for connect-kotlin

### PR DESCRIPTION
**Pull request by buildbreaker**
# Connect GET protocol (Unary only)
Set a `rpc`'s option `idempotency_level` to `NO_SIDE_EFFECTS`. The generated code will pass that through to the Connect runtime.
```proto
  rpc CacheableUnaryCall(SimpleRequest) returns (SimpleResponse) {
    option idempotency_level = NO_SIDE_EFFECTS;
  }
```

# The Connect GET Protocol:
The url will be encoded with the request payload as a query parameter. An example of a url could look like the following:
`https://connect.build/buf.connect.example.Service/CacheableUnaryCall?base64=1&connect=v1&encoding=json&message=...`
The exhaustive list of what parameter keys used:
| **key name**| **example value(s)**| **description**|
|-----|----|---|
| `base64` | `1` or `0` | `1` if the payload is base64 encoded |
| `compression`| `gzip` | The compression used for the payload |
| `connect` | `v1` | The Connect protocol version|
| `message` | binary-data | The encoded payload string |

### CODE REVIEW

File: extensions/google-java/src/main/kotlin/build/buf/connect/extensions/GoogleJavaJSONAdapter.kt [hunk 1]
```diff
@@ -56,7 +56,14 @@ internal class GoogleJavaJSONAdapter<E : GeneratedMessageV3>(
...
     override fun serialize(message: E): Buffer {
-        val jsonString = JsonFormat.printer().print(message)
+        val jsonString = JsonFormat.printer().print(message as E)
+        return Buffer().write(jsonString.encodeUtf8())
+    }
+
+    override fun deterministicSerialize(message: E): Buffer {
+        val jsonString = JsonFormat.printer()
+            .sortingMapKeys()
+            .print(message)
```
Comment (rebello95) [inline extensions/google-java/src/main/kotlin/build/buf/connect/extensions/GoogleJavaJSONAdapter.kt:66]:
```suggestion
            .print(message as E)
```
?

File: extensions/google-java/src/main/kotlin/build/buf/connect/extensions/GoogleJavaProtoAdapter.kt [hunk 2]
```diff
@@ -52,6 +54,20 @@ internal class GoogleJavaProtoAdapter<E : GeneratedMessageV3>(
     }
 
     override fun serialize(message: E): Buffer {
-        return Buffer().write(message.toByteArray())
+        val casted = message
```
Comment (rebello95) [inline extensions/google-java/src/main/kotlin/build/buf/connect/extensions/GoogleJavaProtoAdapter.kt:57]: Why is this needed?

File: library/src/main/kotlin/build/buf/connect/Codec.kt [hunk 3]
```diff
@@ -39,6 +39,15 @@ interface Codec<E> {
      */
     fun serialize(message: E): Buffer
 
+    /**
+     * Deterministic Serialization of the input message.
```
Comment (rebello95) [inline library/src/main/kotlin/build/buf/connect/Codec.kt:43]:
```suggestion
     * Deterministic serialization of the input message.
```
May be worth mentioning why this is useful

File: library/src/main/kotlin/build/buf/connect/Idempotency.kt [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+package build.buf.connect
+
+enum class Idempotency {
+    /**
+     * IdempotencyUnknown is the default idempotency level. A procedure with
+     * this idempotency level may not be idempotent. This is appropriate for
+     * any kind of procedure.
+     */
+    IDEMPOTENCY_UNKNOWN,
```
Comment (rebello95) [inline library/src/main/kotlin/build/buf/connect/Idempotency.kt:23]: Just `UNKNOWN`?

File: library/src/main/kotlin/build/buf/connect/MethodSpec.kt [hunk 5]
```diff
@@ -16,16 +16,25 @@ package build.buf.connect
...
+
 /**
  * Represents the minimum set of information to execute an RPC method.
  * Primarily used in generated code.
  *
  * @param path The path of the request.
  * @param requestClass The Kotlin Class for the request message.
  * @param responseClass The Kotlin Class for the response message.
+ * @param idempotency The declared idempotency of a method.
+ * @param method The http method of a request.
```
Comment (rebello95) [inline library/src/main/kotlin/build/buf/connect/MethodSpec.kt:32]:
```suggestion
 * @param method The HTTP method of a request.
```

File: library/src/main/kotlin/build/buf/connect/http/HTTPClientInterface.kt [hunk 6]
```diff
@@ -24,26 +24,29 @@ typealias Cancelable = () -> Unit
  * Interface for a client that performs underlying HTTP requests and streams with primitive types.
  */
 interface HTTPClientInterface {
+
     /**
      * Perform a unary HTTP request.
      *
+     * @param method The method to use for the request.
```
Comment (rebello95) [inline library/src/main/kotlin/build/buf/connect/http/HTTPClientInterface.kt:31]: May be worth clarifying with an example since this is just a string

File: library/src/main/kotlin/build/buf/connect/impl/ProtocolClient.kt [hunk 7]
```diff
@@ -52,16 +53,22 @@ class ProtocolClient(
         val serializationStrategy = config.serializationStrategy
         val requestCodec = serializationStrategy.codec(methodSpec.requestClass)
         try {
+            val requestMessage = if (config.getConfiguration != GETConfiguration.GETDisabled) {
+                // Use deterministic serialization when GET request configuration is set.
+                requestCodec.deterministicSerialize(request)
+            } else {
+                requestCodec.serialize(request)
```
Comment (rebello95) [inline library/src/main/kotlin/build/buf/connect/impl/ProtocolClient.kt:60]: Is there a particular reason why we wouldn't just always deterministically serialize?

File: library/src/main/kotlin/build/buf/connect/protocols/ConnectInterceptor.kt [hunk 8]
```diff
@@ -60,25 +66,46 @@ internal class ConnectInterceptor(
...
+                }
+                val finalRequestBody = if (requestCompression?.shouldCompress(requestMessage) == true) {
+                    requestHeaders.put(CONTENT_ENCODING, listOf(requestCompression.compressionPool.name()))
+                    requestCompression.compressionPool.compress(requestMessage)
+                } else {
+                    requestMessage
+                }
+                val serializationStrategy = clientConfig.serializationStrategy
+                val requestCodec = serializationStrategy.codec(request.methodSpec.requestClass)
+                if (shouldUseGetMethod(request, finalRequestBody)) {
```
Comment (rebello95) [inline library/src/main/kotlin/build/buf/connect/protocols/ConnectInterceptor.kt:81]: Might be nice to break some of this into another function

File: library/src/main/kotlin/build/buf/connect/protocols/ConnectInterceptor.kt [hunk 9]
```diff
@@ -106,6 +133,12 @@ internal class ConnectInterceptor(
         )
     }
 
+    private fun shouldUseGetMethod(request: HTTPRequest, finalRequestBody: Buffer): Boolean {
+        val getConfiguration = clientConfig.getConfiguration
```
Comment (rebello95) [inline library/src/main/kotlin/build/buf/connect/protocols/ConnectInterceptor.kt:137]: inline this below?

File: library/src/main/kotlin/build/buf/connect/protocols/ConnectInterceptor.kt [hunk 10]
```diff
@@ -256,3 +290,26 @@ private fun Headers.toTrailers(): Trailers {
...
+    if (requestCompression?.shouldCompress(payload) == true) {
+        params.add("${GETConstants.COMPRESSION_QUERY_PARAM_KEY}=${requestCompression.compressionPool.name()}")
+    }
+    params.add("${GETConstants.MESSAGE_QUERY_PARAM_KEY}=${payload.readByteString().base64Url()}")
+    params.add("${GETConstants.BASE64_QUERY_PARAM_KEY}=1")
+    params.add("${GETConstants.ENCODING_QUERY_PARAM_KEY}=${codec.encodingName()}")
+    params.add("${GETConstants.CONNECT_VERSION_QUERY_PARAM_KEY}=${GETConstants.CONNECT_VERSION_QUERY_PARAM_VALUE}")
+    params.sort()
+    val queryParams = params.joinToString("&")
+    val host = baseURL.toURI()
```
Comment (rebello95) [inline library/src/main/kotlin/build/buf/connect/protocols/ConnectInterceptor.kt:312]: This is not the host

File: library/src/main/kotlin/build/buf/connect/protocols/GETConfiguration.kt [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buf.connect.protocols
+
+import okio.Buffer
+
+object GETConstants {
```
Comment (rebello95) [inline library/src/main/kotlin/build/buf/connect/protocols/GETConfiguration.kt:19]: Internal?

File: library/src/main/kotlin/build/buf/connect/protocols/GETConfiguration.kt [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+        override fun useGET(buffer: Buffer): Boolean {
+            return maxUrlBytes > buffer.size
+        }
+    }
+
+    object GETWithoutFallback : GETConfiguration() {
+        override fun useGET(buffer: Buffer): Boolean {
+            return true
+        }
+    }
```
Comment (rebello95) [inline library/src/main/kotlin/build/buf/connect/protocols/GETConfiguration.kt:48]: Not sure what the best practice in Kotlin would be for this, but in Swift I'd do:

- `disabled`
- `enabled`
- `enabledWithFallback(maxURLBytes: Int)`

File: library/src/main/kotlin/build/buf/connect/Idempotency.kt (hunk 4 above)
Comment (buildbreaker) [reply library/src/main/kotlin/build/buf/connect/Idempotency.kt:23]: Can. I was going off of: https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/descriptor.proto#L813C19-L817.

Comment (rebello95) [reply library/src/main/kotlin/build/buf/connect/Idempotency.kt:23]: Yeah I think we can bias towards language best practices in our interfaces

File: library/src/main/kotlin/build/buf/connect/impl/ProtocolClient.kt (hunk 7 above)
Comment (buildbreaker) [reply library/src/main/kotlin/build/buf/connect/impl/ProtocolClient.kt:60]: Performance concerns primarily

**Review by rebello95 — APPROVED** [review]
